### PR TITLE
Stabilize or_patterns (RFC 2535, 2530, 2175)

### LINKED ISSUE #54883: Tracking issue for RFC 2535, 2530, 2175, "Or patterns, i.e `Foo(Bar(x) | Baz(x))`"
This is a tracking issue for the RFC "Or patterns, i.e `Foo(Bar(x) | Baz(x))`" (rust-lang/rfcs#2535).

This issue also tracks the changes in rust-lang/rfcs#2175 and rust-lang/rfcs#2530 since RFC 2535 subsume those.

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [ ] Polish the implementation
- [x] Dogfood or-patterns in the compiler and standard library (done in #71220, #71221, and #71231)
- [ ] Accumulate good test coverage
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- [ ] Should we allow `top_pat` or `pat<allow_top_alt>` in `inferrable_param` such that closures permit `|Ok(x) | Err(x)|` without first wrapping in parenthesis?

   We defer this decision to stabilization as it may depend on experimentation. Our current inclination is to keep the RFC as-is because the ambiguity is not just for the compiler; for humans, it is likely also ambiguous and thus harder to read.

   This also applies to functions which, although do not look as ambiguous, benefit from better consistency with closures. With respect to function arguments there's also the issue that not disambiguating with parenthesis makes it less clear whether the type ascription applies to the or-pattern as a whole or just the last alternative.

- [ ] Should the `pat` macro fragment specifier match `top_pat` in different
   Rust editions or should it match `pat<no_top_alt>` as currently specified? We defer such decisions to stabilization because it depends on the outcome of crater runs to see what the extent of the breakage would be.

The benefit of avoiding `pat<no_top_alt>` in as many places as possible would both be grammatical consistency and fewer surprises for uses. The drawbacks would be possible ambiguity or backtracking for closures and breakage for macros.

**Implementation history:**

- On 2019-08-18, basic parsing landed in https://github.com/rust-lang/rust/pull/61708 landed which was written by @dlrobertson and reviewed by @varkor, @Centril, @petrochenkov, @matthewjasper, and @alexreg.

- On 2019-08-27, full and polished parsing landed in https://github.com/rust-lang/rust/pull/63693 which was written by @Centril and reviewed by @estebank.

- On 2019-08-29, a typo in #63693 was fixed in https://github.com/rust-lang/rust/pull/63938 which was written by @tshepang and reviewed by @Centril.

- On 2019-09-05, https://github.com/rust-lang/rust/pull/64128, written by @Centril and reviewed by @davidtwco, landed. The PR fixed issue https://github.com/rust-lang/rust/issues/64106 where the compiler would spuriously emit `unused_parens` warnings on:

    - `ref? mut? x @ (p_0 | ... | p_n)`
    - `box (p_0 | ... | p_n)`
    - `& mut? (p_0 | ... | p_n)`
    - `|(p_0 | ... | p_n)| $expr`
    - `fn foo((p_0 | ... | p_n): $type)`

- On 2019-09-06, https://github.com/rust-lang/rust/pull/64111, written by @Centril and reviewed by @petrochenkov, landed. The PR fixed the late-resolution behavior of or-patterns in nested positions including the already-bound check and binding-consistency check. Moreover, the AST now uniformly uses `ast::PatKind::Or` and has no special means of representing or-patterns at the top level.

- On 2019-09-25, https://github.com/rust-lang/rust/pull/64508, written by @Centril and reviewed by @matthewjasper, @varkor, and @nikomatsakis, landed. The PR adjusted the HIR and HAIR to consistently use or-patterns at the top & nested levels. Moreover, liveness, typeck, and dead_code analyses were adjusted also. Notably, The PR did however not adjust exhaustiveness checking. Tangentially, https://github.com/rust-lang/rust/pull/64271, landed 2019-09-12, written by @Centril and reviewed by @estebank, adjusted `check_match` diagnostics logic to be better prepared for or-patterns.

- On 2019-09-29, https://github.com/rust-lang/rust/pull/64887, written by @Centril and reviewed by @estebank, landed. The PR improved parser recovery for trailing `|` in pattern contexts.

- On 2019-10-27, https://github.com/rust-lang/rust/pull/65874, written by @Nadrieril and reviewed by @varkor, @arielb1, @nnethercote, and @Centril, landed. The PR refactored the usefulness/exhaustiveness algorithm in match checking. This laid the foundation upon which exhaustiveness checking for or-patterns was later initially implemented in https://github.com/rust-lang/rust/pull/66612.

- On 2019-11-23, https://github.com/rust-lang/rust/pull/66639, written by @Centril and reviewed by @petrochenkov, landed. In this follow-up to https://github.com/rust-lang/rust/pull/64111, the resolve logic for adding fresh bindings was simplified and re-written in a more declarative fashion.

- On 2019-12-01, https://github.com/rust-lang/rust/pull/66612, written by @Nadrieril and reviewed by @varkor, @Centril, and @matthewjasper, landed. In this follow-up to https://github.com/rust-lang/rust/pull/65874, an initial implementation of exhaustiveness / usefulness checking for or-patterns was implemented.

- On 2019-12-03, https://github.com/rust-lang/rust/pull/66967, written by @Nadrieril and reviewed by @varkor and @Centril, landed. In this follow-up to https://github.com/rust-lang/rust/pull/66612, the top-level hack for or-patterns in exhaustiveness checking was removed, making or-patterns truly first-class in match checking.

- On 2019-12-21, https://github.com/rust-lang/rust/pull/67428, written by @Centril and reviewed by @matthewjasper, landed. This PR adjusted regionck's `resolve_local::is_binding_pat` such that the `P&` grammar includes or-patterns.

- On 2019-12-22, https://github.com/rust-lang/rust/pull/67439, written by @Centril and reviewed by @matthewjasper, landed. Among other things, this PR simplified HAIR lowering of or-patterns a smidgen.

- On 2019-12-26, https://github.com/rust-lang/rust/pull/67592, written by @matthewjasper and reviewed by @Centril, landed. This PR did some preparatory cleanups in MIR lowering of `match` expressions before the MIR level support for or-patterns is added. Moreover, the PR gated or-patterns in const contexts under `const_if_match`.

- On 2020-02-04, https://github.com/rust-lang/rust/pull/67668, written by @matthewjasper and reviewed by @pnkfelix, @Centril, @nagisa, @varkor, and @eddyb, landed. The PR finally added the support for or-patterns in MIR building (dynamic semantics, borrow checking, ...). Now, nested or-patterns can actually be used on nightly with `#![feature(or_patterns)]`.

- On 2020-02-06, https://github.com/rust-lang/rust/pull/68842, written by @Centril and reviewed by @estebank landed. The PR added a regression test for an issue fixed by https://github.com/rust-lang/rust/pull/67668.

- On 2020-02-15, https://github.com/rust-lang/rust/pull/68856, written by @Centril and reviewed by @matthewjasper landed. The PR polished the type checking of patterns with respect to default binding modes. In relation to or-patterns, the PR declared that or-patterns are pass-through (`AdjustMode::Pass`) in terms of default binding modes. This was already the case in the implementation (and already stabilized for top level or-patterns) but the PR removed the FIXME that left the question open.

- On 2020-02-28, 2020-03-08, and 2020-03-08, PRs https://github.com/rust-lang/rust/pull/69452, https://github.com/rust-lang/rust/pull/69599, and https://github.com/rust-lang/rust/pull/69687 landed. These were written by @Centril and were reviewed by @estebank, @davidtwco, and @estebank respectively. All PRs tweaked and improved diagnostics wrt. or-patterns.

- On 2020-03-08, https://github.com/rust-lang/rust/pull/69690, written by @thekuom and reviewed by @Centril, landed. The relevant part of the PR added tests for the dynamic semantics of or-patterns combined with `x @ y` patterns, `box` patterns, and slice patterns.

- On 2020-03-10, https://github.com/rust-lang/rust/pull/69817, written by @thekuom and reviewed by @Centril, landed. The PR added borrow checker tests in combination with other pattern features.

- On 2020-03-11, https://github.com/rust-lang/rust/pull/69891, written by @Centril and reviewed by @varkor and @Nadrieril, landed. The PR fixed a bug (https://github.com/rust-lang/rust/issues/69875) in exhaustiveness checking where the patterns `0 | (1 | 2)` and `x @ 0 | x @ (1 | 2)` would result in an ICE.

- On 2020-03-27, https://github.com/rust-lang/rust/pull/70413, written by @AminArria and reviewed by @varkor, @Nadrieril, and @Centril, landed. The PR fixed https://github.com/rust-lang/rust/issues/70372, a false positive  in the linting behavior of `unreachable_patterns` when combining or-patterns, bindings, and guards.

- On 2020-04-13, https://github.com/rust-lang/rust/pull/67766, written by @sapir and reviewed by @matthewjasper, landed. The PR fixed https://github.com/rust-lang/rust/issues/67691, a bug in the `unused_variables` lint pertaining to or-patterns.

### PR DESCRIPTION
**Pull request by mark-i-m**
closes #54883

This PR stabilizes the or_patterns feature in Rust 1.53.

This is blocked on the following (in order):
- [x] The crater run in https://github.com/rust-lang/rust/pull/78935#issuecomment-731564021
- [x] The resolution of the unresolved questions and a second crater run (https://github.com/rust-lang/rust/pull/78935#issuecomment-735412705)
    - It looks like we will need to pursue some sort of edition-based transition for `:pat`.
- [x] Nomination and discussion by T-lang
- [x] Implement new behavior for `:pat` based on consensus (https://github.com/rust-lang/rust/pull/80100).
- [ ] An FCP on stabilization

EDIT: Stabilization report is in https://github.com/rust-lang/rust/pull/79278#issuecomment-772815177

### CODE REVIEW

**Comment by mark-i-m** [comment]
I'm not sure I understand the CI failure... I thought `cfg_attr(bootstrap...)` wasn't needed in the compiler?

**Comment by est31** [comment]
@mark-i-m it's only not needed when you stabilize a std feature as std is always compiled first.

**Comment by mark-i-m** [comment]
@est31 Do you happen to know how to bless clippy tests? Everything is passing locally on stage 2.

**Comment by est31** [comment]
@mark-i-m sorry no idea off the top of my head. IIRC the compiletest code creates the files on the disk and you only have to copy them to the right location. bless is only a wrapper around that, but no idea whether clippy has better facilities than manual copies.

**Comment by camelid** [comment]
> `@est31` Do you happen to know how to bless clippy tests? Everything is passing locally on stage 2.

Unfortunately you can't bless Clippy tests in-tree :/

I just [asked about this the other day in #clippy](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/blessing.20clippy.20tests.20in-tree).

**Comment by mark-i-m** [comment]
@camelid I see. So what should one do with the CI failure? Does it just stay that way?

**Comment by camelid** [comment]
I think you might need to update by hand :/

**Comment by mark-i-m** [comment]
AFAICT this is waiting on hearing back from a team member and maybe having an FCP.

@rustbot modify labels: -S-blocked +S-waiting-on-review

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting, and we're prepared to start FCP to merge/stabilize as soon as we have a stabilization report.

**Comment by mark-i-m** [comment]
Ah, yeah, I forgot about that. Unfortunately, I have neither the knowledge nor the time to produce the full report atm. I would need help from someone else who knows more  (perhaps @varkor or @Nadrieril ?). 


> The report should contain:
>
>    A summary, showing examples (e.g. code snippets) what is enabled by this feature.
>    Links to test cases in our test suite regarding this feature and describe the feature's behavior on encountering edge cases.
>    Links to the documentations (the PRs we have made in the previous steps).
>    Any other relevant information(Examples of such reports can be found in rust-lang/rust#44494 and rust-lang/rust#28237).
>    The resolutions of any unresolved questions if the stabilization is for an RFC.

I can write about the first and last points, but don't know much about the others...

Should I start a hackmd or something?

**Comment by mark-i-m** [comment]
Started this hackmd: https://hackmd.io/oczPRxyBSKOAojC1Dgxupw

**Comment by nikomatsakis** [comment]
One other note regarding stabilization:

Can we avoid stabilize the pat2018/pat2021 matchers until the edition is stabilized, just in case we slip in the edition and want to update the names?

**Comment by nikomatsakis** [comment]
Also, once this is stabilized, we need to update the grammar in the reference. I'm not sure what protocol we usually follow... I think maybe the best practice is to have a pending PR that gets merged after we accept this one?

(cc @rust-lang/docs)

**Comment by GuillaumeGomez** [comment]
I think this is the best course indeed. Or even open a PR referencing the one before this one is merged. I don't think the order matters much here.

**Comment by joshtriplett** [comment]
> Can we avoid stabilize the pat2018/pat2021 matchers until the edition is stabilized, just in case we slip in the edition and want to update the names?

I don't think it's unreasonable to commit to calling it "Rust 2021", at this point, even if it were to get released in (for instance) early 2022.

**Comment by mark-i-m** [comment]
> Can we avoid stabilize the pat2018/pat2021 matchers until the edition is stabilized, just in case we slip in the edition and want to update the names?

Yes, this PR does _not_ intend to stabilize these.

Sorry for the delay... I'm a bit busy for the next week or so. Then, hopefully I can come back to this.

**Comment by steveklabnik** [comment]
>  I'm not sure what protocol we usually follow... I think maybe the best practice is to have a pending PR that gets merged after we accept this one?

We relaxed this, and the process currently is "file an issue on the reference". PRs fail until the functionality is stable, and the whole "send docs PR, merge impl PR, restart tests for docs PR, merge" dance was too much trouble.

**Comment by mark-i-m** [comment]
I have rebased and submitted https://github.com/rust-lang/reference/pull/957 to update the reference. Since beta is branching soon, I've bumped the version in the PR to 1.52...

Also thanks to @Nadrieril for help with this report:

# Stabilization Report

This is a stabilization report for the `or_patterns` feature. It does _NOT_ stabilize the `edition_macro_pat` feature; i.e. `:pat2018/21` remain unstable.

A mighty Thank You to the many people who designed, implemented, and reviewed this feature, including (in no particular order, and apologies to any I have missed) Centril, varkor, Nadrieril, matthewjasper, petrochenkov, alexreg, estebank, davidtwco, nikomatsakis, arielb1, nnethercote, pnkfelix, nagisa, eddyb, thekuom, sapir, Nemo157, and joshtriplett :tada:

# Examples

This stabilization allows nested or-patterns in stable Rust in all of the following cases:

1. `if let` expressions:

   ```rust
   if let Foo::Bar | Foo::Quux(1 | 2) = some_computation() {
       ...
   }
   ```

1. `while let` expressions:

   ```rust
   while let Ok(1 | 2) | Err(3) = different_computation() {
       ...
   }
   ```

3. `let` statements:

   ```rust
   let Ok(x) | Err(x) = another_computation();
   ```

   In this case, the pattern must be irrefutable as `Ok(x) | Err(x)` is.

4. `fn` arguments:

   ```rust
   fn foo((Ok(x) | Err(x)): Result<u8, u8>) {
       ...
   }
   ```

   Here too, the pattern must be irrefutable.

5. `match` patterns:

   ```rust
   match foo {
       Foo::Bar | Foo::Quux(1 | 2) => { ... }
       ...
   }
   ```

It does _not_ stabilize or-patterns in closure arguments or macro `:pat` matchers (more below).

# Test cases

Test cases can be found in [the dedicated folder](https://github.com/rust-lang/rust/tree/a4cbb44ae2c80545db957763b502dc7f6ea22085/src/test/ui/or-patterns).

Additionally, the following tests test the behavior of `:pat` in different editions:
- [src/test/ui/macros/macro-pat-follow.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/macro-pat-follow.rs)
- [src/test/ui/macros/macro-pat-follow-2018.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/macro-pat-follow-2018.rs)
- [src/test/ui/macros/edition-macro-pats.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/macros/edition-macro-pats.rs)

And the feature gate test for `edition_macro_pat`:
- [src/test/ui/feature-gate-edition_macro_pats.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/feature-gate-edition_macro_pats.rs)

# Documentation

The RFC contains the precise [grammar changes][grammar].

PR to update the reference: https://github.com/rust-lang/reference/pull/957

# Other Relevant Info

petrochekov [raised a concern](https://github.com/rust-lang/rust/issues/54883#issuecomment-742802350) about simplifying the grammar for patterns to allow a leading vert (`|`) at nested levels, not just at the top level. There was [an FCP](https://github.com/rust-lang/rust/issues/81415), and there is team consensus for doing so. However, we do not implement it in this PR because it is a backwards-compatible change. It allows strictly more code to compile. The only place where such a grammar change could be visible is in macros in 2021+, which are still unstable.

# Resolved Unresolved Questions

> Should we allow top_pat or pat<allow_top_alt> in inferrable_param such that closures permit |Ok(x) | Err(x)| without first wrapping in parenthesis?

In the [2020-11-10 T-lang meeting][tlangcon], there was consensus this should not be allowed. Despite the consistency with other places where patterns are allowed, this syntax was deemed to be too confusing.

> Should the pat macro fragment specifier match top_pat in different Rust editions or should it match pat<no_top_alt> as currently specified? We defer such decisions to stabilization because it depends on the outcome of crater runs to see what the extent of the breakage would be.

Originally, it was thought that `:pat` could match `top_pat` in all editions, but two crater runs ([0][crater0], [1][crater1]) showed that even with a fallback plan, there was lots of complex breakage.

Instead, in the [2020-12-15 T-lang meeting][tlangcon2], there was consensus that in Rust 2015/18, `:pat` would continue to match `pat<no_top_alt>` (as it already does), while in Rust 2021 and onward, it would match `top_pat`. This means that in Rust 2015/18, `:pat` will _not_ match top-level or-patterns, but in subsequent editions it will.

Additionally, to allow cross-edition macro usage, the `:pat2018` and `:pat2021` matchers were introduced behind the `edition_macro_pat` feature, which is _not_ being stabilized in this stabilization report.


[grammar]: https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md#reference-level-explanation
[tlangcon]: https://github.com/rust-lang/rust/issues/54883#issuecomment-724898641
[crater0]: https://github.com/rust-lang/rust/pull/78935#issuecomment-735050394
[crater1]: https://github.com/rust-lang/rust/pull/78935#issuecomment-740828151
[tlangcon2]: https://github.com/rust-lang/rust/issues/54883#issuecomment-745509090

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by scottmcm** [comment]
Hmm, I explored a bit and I'd like to bring up a bit around mixing these patterns with type annotations.

The following errors, saying that it needs parens around the pattern:
```rust
fn bar(Ok(x) | Err(x): Result<u8, u8>) {}
```

However the equivalent in a `let` doesn't warn:
```rust
let Ok(x) | Err(x): Result<u8, u8> = y;
```

1. Why are the parens needed in the first example?  I understand them as being desirable for closures (because of `|` being used in both syntaxes), but didn't see a mention of normal functions.  (Of course it's completely possible I just missed it.)
2. The RFC describes `|` in patterns as having "low precedence".  Is there any guarding that would be worth doing here in case of adding more ascription to patterns?  For example, we could allow `let Ok(x) | Err(x) = y;` but require parenthesizing with a type annotation like `let (Ok(x) | Err(x)): Result<u8, u8> = y;`.  My vague expectation would be that, if the annotations were part of the pattern grammar instead of the `let` grammar, `Ok(x) | Err(x): Result<u8, u8>` would parse as `Ok(x) | (Err(x): Result<u8, u8>)`, and **not** as `(Ok(x) | Err(x)): Result<u8, u8>`.  But maybe that difference would be irrelevant because all parts of an or pattern need to have the same type?

**Comment by mark-i-m** [comment]
@scottmcm

1. The RFC says that fn args take `pat<no_top_alt>` which requires the parens.

2. I think you are correct. If we want future compatibility with nested type ascription, we may want to change let to also use  `pat<no_top_alt>`. I think none of the others need to change though.

**Comment by petrochenkov** [comment]
FWIW, we can always keep the combination "or pattern" + "type ascription" unstable  to unblock or patterns in general and answer the question about priorities later.

**Comment by Mark-Simulacrum** [comment]
> It does not stabilize or-patterns in closure arguments or macro :pat matchers (more below).

Am I correct in assuming that this stabilizes or patterns in all places a pattern can go *except* for these two that are specifically called out? I see a listing in the stabilization report, but since there's not language specifically saying "all except for" I'm worried perhaps I overlooked something.

**Comment by mark-i-m** [comment]
@Mark-Simulacrum Yes

**Comment by mark-i-m** [comment]
Working on https://github.com/rust-lang/rust/pull/82048 to disallow or-patterns before type-ascription, as per @petrochenkov's suggestion.

**Comment by scottmcm** [comment]
@rfcbot concern parens-in-let

We [discussed](https://hackmd.io/7f-Y8vrKSkaIecxahDemqQ?view#%E2%80%9CStabilize-or_patterns-RFC-2535-2530-2175%E2%80%9D-rust79278) this in this week's @rust-lang/lang team meeting.  We would like to see all three annotated positions -- function parameters, closure parameters, and lets -- all require the parentheses for now.  (Thus `pat<no_top_alt>` for all of them, like you mentioned in https://github.com/rust-lang/rust/pull/79278#issuecomment-773371179, as I understand it.)

There was not agreement on whether this should be the final state -- specifically some thought that no parens should be needed for the non-annotated case, while others preferred them there too -- but the hope was that using the same grammar production in the three places for now would let the feature stabilize without being too annoying to implement, and then we'll learn how annoying the parens are in practice as more people start using this.

**Comment by mark-i-m** [comment]
@scottmcm https://github.com/rust-lang/rust/pull/82048 prohibits top-level or-patterns in `let` bindings, as per you comment. I've also updated the [stabilization report](https://github.com/rust-lang/rust/pull/79278#issuecomment-772815177).

**Comment by mark-i-m** [comment]
Should this be renominated? It looks like we still need a few more boxes checked.

**Comment by joshtriplett** [comment]
cc @cramertj @nikomatsakis @pnkfelix

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by mark-i-m** [comment]
Rebased.

Also, I changed the stabilization version to 1.53, since the beta will branch before the FCP finishes.

File: compiler/rustc_expand/src/expand.rs [hunk 1]
```diff
@@ -917,7 +917,7 @@ pub fn parse_ast_fragment<'a>(
         }
         AstFragmentKind::Ty => AstFragment::Ty(this.parse_ty()?),
         AstFragmentKind::Pat => {
-            AstFragment::Pat(this.parse_pat_allow_top_alt(None, GateOr::Yes, RecoverComma::No)?)
+            AstFragment::Pat(this.parse_pat_allow_top_alt(None, RecoverComma::No)?)
```
Comment (nikomatsakis) [inline compiler/rustc_expand/src/expand.rs:920]: I'm a bit confused about this =) does this mean that `pat` accepts `a|b`? I guess I expected some edition logic, and the new fragment matchers, is that in another PR?

Comment (mark-i-m) [reply compiler/rustc_expand/src/expand.rs:920]: I didn't fully understand this change very well either, but @petrochenkov seems confident (https://github.com/rust-lang/rust/pull/81869#discussion_r573234448). Apparently, this code is for parsing a pattern produced by a macro expansion.

The code you are probably thinking of is here: https://github.com/rust-lang/rust/pull/79278/files#diff-1a054923fe6e121f46954f52f6660a22970818b5c8b3ef1c0f3b703eae2baef6L122-L128

**Pull request merged**